File: tuner/src/tuner_cli/statistics.py
"""Pure pair-level scoring and conservative comparison rules."""

from __future__ import annotations

import math
import random

from .domain import Estimate, GameResult, PairResult

ALPHA = 0.05


def game_utility(game: GameResult) -> float:
    return {"candidate_win": 1.0, "draw": 0.5, "baseline_win": 0.0}[game.outcome]


def pair_utility(pair: PairResult) -> float:
    return sum(game_utility(game) for game in pair.games) / 2


def marginal_interval(values: tuple[float, ...]) -> Estimate:
    if not values:
        raise ValueError("an interval needs at least one pair")
    mean = sum(values) / len(values)
    half = math.sqrt(math.log(2 / ALPHA) / (2 * len(values)))
    return Estimate(mean, max(0.0, mean - half), min(1.0, mean + half))


def paired_difference_values(left: tuple[float, ...], right: tuple[float, ...]) -> Estimate:
    if not left or len(left) != len(right):
        raise ValueError("paired differences need equal non-empty inputs")
    mean = sum(a - b for a, b in zip(left, right, strict=True)) / len(left)
    half = math.sqrt(2 * math.log(2 / ALPHA) / len(left))
    return Estimate(mean, max(-1.0, mean - half), min(1.0, mean + half))


def paired_difference(left: tuple[float, ...], right: tuple[float, ...]) -> Estimate:
    """Compatibility numeric primitive; contextual callers use observations.paired_difference."""
    return paired_difference_values(left, right)


def tie_relation(difference: Estimate) -> str:
    if difference.lower > 0:
        return "better"
    if difference.upper < 0:
        return "worse"
    return "tie"


def bootstrap_mean_interval(
    values: tuple[float, ...], seed: int, resamples: int = 4096
) -> Estimate:
    """Deterministic percentile bootstrap for independent complete runs."""
    if len(values) < 2 or any(not math.isfinite(value) for value in values) or resamples <= 0:
        raise ValueError("bootstrap needs two finite values and positive resamples")
    rng = random.Random(seed)
    means = sorted(
        sum(values[rng.randrange(len(values))] for _ in values) / len(values)
        for _ in range(resamples)
    )
    # Nearest-rank order statistic keeps the encoded interval stable across Python versions.
    return Estimate(
        sum(values) / len(values),
        means[round(0.025 * (resamples - 1))],
        means[round(0.975 * (resamples - 1))],
    )

File: tuner/domain-model/src/Statistics.hs
module Statistics where

-- | The fixed Hoeffding confidence level used across the tuner.
alpha :: Double
alpha = 0.05

-- | Game outcome utility: win=1, draw=0.5, loss=0.
data Utility = Utility Double
  deriving (Eq, Show)

-- | A point estimate with Hoeffding confidence bounds.
data Estimate = Estimate
  { estMean  :: Double
  , estLower :: Double
  , estUpper :: Double
  }
  deriving (Eq, Show)

-- | Game utility from an outcome string.
gameUtility :: String -> Double
gameUtility = undefined

-- | Pair utility: average of the two games (seat-balanced, [0,1]).
pairUtility :: Double -> Double -> Double
pairUtility first second = (first + second) / 2.0

-- | Marginal Hoeffding interval (alpha = 0.05) for a sequence of pair utilities.
marginalInterval :: [Double] -> Estimate
marginalInterval = undefined

-- | Numeric primitive for paired difference intervals between two
-- equal-length sequences; contextual callers use 'Observations.pairedDifference'.
pairedDifferenceValues :: [Double] -> [Double] -> Estimate
pairedDifferenceValues = undefined

-- | Deterministic percentile bootstrap interval for independent complete runs.
bootstrapMeanInterval :: [Double] -> Int -> Int -> Estimate
bootstrapMeanInterval = undefined

-- | Tie relation from a difference estimate.
data TieRelation = Better | Worse | Tie
  deriving (Eq, Show)

tieRelation :: Estimate -> TieRelation
tieRelation = undefined